core/src/main/scala/akka/kafka/internal/ProducerStage.scala

```diff
@@ -6,126 +6,284 @@
-            producer.flush()
-            producer.close(closeTimeout.toMillis, TimeUnit.MILLISECONDS)
-            log.debug("Producer closed")
-          }
-          catch {
-            case NonFatal(ex) => log.error(ex, "Problem occurred during producer close")
-          }
+    private def onCommitInterval(): Unit = {
+      if (!commitInProgress) {
+        commitInProgress = true
+      }
+      val awaitingConfirmationCount = awaitingConfirmation.get
+      if (awaitingConfirmationCount == 0) {
+        maybeCommitTransaction()
+        commitInProgress = false
+        // restart demand for more messages once last batch has been committed
+        if (!hasBeenPulled(in)) {
+          tryPull(in)
         }
+        scheduleOnce(commitSchedulerKey, commitIntervalMs.milliseconds)
+      }
+      else {
+        scheduleOnce(commitSchedulerKey, messageDrainIntervalMs.milliseconds)
+      }
+    }
 
-        super.postStop()
+    override val onMessageAckCb: AsyncCallback[Message[K, V, P]] = getAsyncCallback[Message[K, V, P]] { (msg) =>
+      msg.passThrough match {
+        case o: ConsumerMessage.PartitionOffset => batchOffsets.update(o)
+        case _ =>
       }
     }
-    logic
+
+    override def onCompletionSuccess(): Unit = {
+      log.debug("Committing final transaction before shutdown")
+      maybeCommitTransaction(beginNewTransaction = false)
+    }
+
+    override def onCompletionFailure(ex: Throwable): Unit = {
+      log.debug("Aborting transaction due to stage failure")
+      abortTransaction()
+    }
+
+    private def maybeCommitTransaction(beginNewTransaction: Boolean = true): Unit = {
+      if (batchOffsets.nonEmpty) {
+        val (groupId, offsetMap) = batchOffsets.offsetMap()
+        log.debug(s"Committing transaction for consumer group '${groupId}' with offsets: $offsetMap")
+        producer.sendOffsetsToTransaction(offsetMap, groupId)
+        producer.commitTransaction()
+        batchOffsets.clear()
+        isTransactionOpen = false
+        if (beginNewTransaction) {
+          beginTransaction()
+        }
+      }
+    }
+
+    private def initTransactions(): Unit = {
+      log.debug("Initializing transactions")
+      producer.initTransactions()
+    }
+
+    private def beginTransaction(): Unit = {
+      log.debug("Beginning new transaction")
+      producer.beginTransaction()
+      isTransactionOpen = true
+    }
+
+    private def abortTransaction(): Unit = {
+      log.debug("Aborting transaction")
+      producer.abortTransaction()
+      isTransactionOpen = false
+    }
+  }
+}
+
+private[kafka] trait ProducerCompletionState {
+  def onCompletionSuccess(): Unit = {}
+  def onCompletionFailure(ex: Throwable): Unit = {}
+}
+
+private[kafka] trait MessageCallback[K, V, P] {
+  def awaitingConfirmation: AtomicInteger
+  def onMessageAckCb: AsyncCallback[Message[K, V, P]]
+}
+
+private[kafka] class TransactionOffsetBatch() {
+  private val offsets = mutable.Map[GroupTopicPartition, Long]()
```

Comment:
We don't need to be concerned with concurrent modifications because this Map is updated from the `GraphStateLogic` which ensures threadsafety for us.  We could still make `TransactionOffsetBatch` immutable, but we effectively just push that mutation down `TransactionProducerStageLogic` which may be a better place for it anyway.  I'll consider this some more.